### DISCUSSION TITLE
`gopls` not working as intended from `lsp`

### DISCUSSION THREAD

**Discussion by mooyg** [nix-community/nixvim#1442] [Q&A] [answered]
`gopls` doesn't work as expected it only complains about the syntax errors and nothing else in a file. I wonder this might be due to it not being perfectly configured? because I have this which works fine 
```go
   extraConfigLua = ''
      local lspconfig = require("lspconfig")
      lspconfig.gopls.setup({
        settings = {
          gopls = {
            analyses = {
              unusedparams = true,
            },
            staticcheck = true,
            gofumpt = true,
          },
        },
      })    '';
```
if I just use the `gopls` from the `lsp` it initialises but doesn't work as intended, the Client is also attached to the Buffer and no errors at all.

**Comment by GaetanLepage** [comment]
Is there a reason to configure it in lua directly ?
Have you tried to simply set `plugins.lsp.servers.gopls.enable` to `true` ?

**Reply by mooyg** [reply]
yes, I have and it doesn't really works how you expect it too. Configuring with lua works for now. Just enabling it from plugins attaches it to the Buffer but doesn't show anything expect syntax errors.

**Reply by GaetanLepage** [reply]
What are those missing features ? Go to definition, documentation etc ?
If you use the lua snippet just so that you can provide `settings`, know that you can also do it from nix:
```nix
plugins.lsp.servers.gopls = {
  enable = true;
  extraOptions.settings.gopls = {
    analyses.unusedparams = true;
    staticcheck = true;
    gofumpt = true;
  };
};
```

**Reply by GaetanLepage** [reply]
By inspecting the generated `init.lua` (`nixvim-print-init`), you will be able to check that the produced lua snippet should be functionally the same as the one you provided.

**Reply by mooyg** [reply]
This doesn't work 
![CleanShot 2024-04-23 at 15 01 08@2x](https://github.com/nix-community/nixvim/assets/61348580/3c5f4403-882a-498e-af5b-e23eebe923a1)
using nix config no error is returned here
and same using the extra config 
![CleanShot 2024-04-23 at 15 01 37@2x](https://github.com/nix-community/nixvim/assets/61348580/69d58a10-4eeb-4822-8af0-8815373c65a3)

```lua
  local __lspServers = {{["name"] = "yamlls"},{["name"] = "volar"},{["name"] = "tsserver"},{["name"] = "taplo"},{["name"] = "tailwindcss"},{["name"] = "svelte"},{["name"] = "rust_analyzer"},{["name"] = "prismals"},{["name"] = "nil_ls"},{["name"] = "marksman"},{["extraOptions"] = {["cmd"] = {"/nix/store/r6a2hwab6bsw81z0ybsp0q7cv9qnlp73-vscode-langservers-extracted-4.8.0/bin/vscode-json-language-server","--stdio"}},["name"] = "jsonls"},{["extraOptions"] = {["settings"] = {["gopls"] = {["analyses"] = {["unusedparams"] = true},["gofumpt"] = true,["staticcheck"] = true}}},["name"] = "gopls"}}
```
init.lua for the following

**Reply by GaetanLepage** [reply]
Ok, so what I suspect is that it is not the same `gopls` binary that is run in both cases.
- When you set `plugins.lsp.servers.gopls.enable = true`, then nixvim installs `pkgs.gopls` for you.
- However, when you simply configure the LS using raw lua, neovim will run whatever `gopls` binary is in your `$PATH` at this time.

So in this second case, maybe you are using the one from your project or similar. They might not have access to the same scope...
However, I don't know a lot about go, so I am not sure on how the LS is handling the project context.

**Reply by mooyg** [reply]
> Ok, so what I suspect is that it is not the same `gopls` binary that is run in both cases.
> 
> - When you set `plugins.lsp.servers.gopls.enable = true`, then nixvim installs `pkgs.gopls` for you.
> 
> - However, when you simply configure the LS using raw lua, neovim will run whatever `gopls` binary is in your `$PATH` at this time.
> 
> 
> 
> So in this second case, maybe you are using the one from your project or similar. They might not have access to the same scope...
> 
> However, I don't know a lot about go, so I am not sure on how the LS is handling the project context.

That could be true indeed, but the weird thing is how could it act differently? Same thing in vscode works too, they use gopls as the language server too so shouldnt really matter where its coming from

**Reply by GaetanLepage** [reply]
Well, I don't know how the go language server from your PATH is actually installed, but it could maybe be configured in another way.
To be honest, I think that this is a specific go issue.
The only thing that we do on our side is to add `pkgs.gopls` to your system packages.
If, for whatever reason, this binary does not integrate well with your code, I cannot really help you.
This is why we have those package options that can be set to `null` to prevent the installation of the tool in question. Like so, you have full control over what language server you use.

I will thus mark this discussion as "closed".
Of course, if you feel like this should be investigated further or that this is actually a nixvim issue, then do not hesitate to open an issue :)

**Reply by GaetanLepage** [reply, marked as answer]
So, to enable the `gopls` language server in nixvim but prevent the installation of `gopls`, do the following:
```nix
plugins.lsp = {
  enable = true;
  servers.gopls = {
    enable = true;
    package = null;
  };
};
```
